### PR TITLE
[smoke-test] Adding multiple chunk sync for state sync smoke test

### PR DESCRIPTION
**Pull request by sunmilee**
### Motivation

For state synchronization, we want to test both single chunk sync and multiple chunk sync. However with prod sized chunks, we need to submit thousands of transactions to test multiple chunk sync. To avoid doing this, we override the config in smoke test to have chunk_limit =2 and test both cases.

This PR also refactors smoke test and swarm to take in NodeConfig rather than filepaths. 

## Test Plan

cargo test -- test_basic_state_synchronization

### CODE REVIEW

**Comment by sunmilee** [comment]
@davidiw there's duplicate code but is this in line with what you were suggesting?

File: testsuite/tests/libratest/smoke_test.rs [hunk 1]
```diff
@@ -160,13 +160,25 @@ impl TestEnvironment {
 fn setup_swarm_and_client_proxy(
     num_nodes: usize,
     client_port_index: usize,
+    template: Option<NodeConfig>,
```
Comment (phoenix-antigravity) [inline testsuite/tests/libratest/smoke_test.rs]: is there any other use case for this parameter?
if it's just to test small chunk size, I'd actually make it default behavior, and keep previous function signature

Comment (sunmilee) [reply testsuite/tests/libratest/smoke_test.rs]: Yep right now that's the only use case. I wonder in the future people might want to make custom environments to run tests on though?

Comment (phoenix-antigravity) [reply testsuite/tests/libratest/smoke_test.rs]: I'd prefer not to optimize for unknowns. We can add it later. Making this parameter as default actually will automatically cover all tests when StateSynchronizer is involved

File: libra-swarm/src/main.rs [hunk 2]
```diff
@@ -40,21 +40,21 @@ fn main() {
...
     .expect("Failed to configure validator swarm");
 
     let mut full_node_swarm = if num_full_nodes > 0 {
         Some(
             LibraSwarm::configure_swarm(
+                None, /* template config */
                 num_full_nodes,
                 RoleType::FullNode,
                 None, /* config dir */
-                None, /* template_path */
```
Comment (davidiw) [inline libra-swarm/src/main.rs]: why did we move the position of the parameter? typically optional parameters are placed at the end.

File: libra-swarm/src/swarm.rs [hunk 3]
```diff
@@ -242,11 +242,17 @@ impl LibraSwarm {
         for i in 0..num_launch_attempts {
             info!("Launch swarm attempt: {} of {}", i, num_launch_attempts);
 
+            let template = if let Some(template_path) = template_path.clone() {
+                NodeConfig::load(workspace_builder::workspace_root().join(template_path))
+                    .unwrap_or_default()
+            } else {
+                NodeConfig::default()
+            };
```
Comment (davidiw) [inline libra-swarm/src/swarm.rs]: I thought we were gonna change the interface here to take in the node config and not the file?

File: libra-swarm/src/swarm.rs [hunk 4]
```diff
@@ -287,28 +293,27 @@ impl LibraSwarm {
...
-            template.vm_config.publishing_options = VMPublishingOption::Open;
+
+        let mut node_config = if let Some(template) = template {
             template
+        } else {
+            NodeConfig::default()
         };
-        template.base.role = role;
+        node_config.vm_config.publishing_options = VMPublishingOption::Open;
+        node_config.base.role = role;
```
Comment (davidiw) [inline libra-swarm/src/swarm.rs]: we should actually move this to the only test that needs it....

File: testsuite/tests/libratest/smoke_test.rs (hunk 1 above)
Comment (davidiw) [reply testsuite/tests/libratest/smoke_test.rs]: There's a good use case above where we can use this, right now we're internally setting weird defaults for smoke tests that aren't there for production, namely the ability to publish.

File: testsuite/tests/libratest/smoke_test.rs [hunk 5]
```diff
@@ -160,13 +160,25 @@ impl TestEnvironment {
...
 ) -> (TestEnvironment, ClientProxy) {
-    let mut env = TestEnvironment::new(num_nodes);
+    let mut env = TestEnvironment::new_with_template(num_nodes, template);
```
Comment (davidiw) [inline testsuite/tests/libratest/smoke_test.rs]: no need to rename this new_with_template, imo.

File: testsuite/tests/libratest/smoke_test.rs [hunk 6]
```diff
@@ -160,13 +160,25 @@ impl TestEnvironment {
...
     env.launch_swarm(RoleType::Validator);
     let ac_client = env.get_validator_ac_client(client_port_index, None);
     (env, ac_client)
 }
 
+//fn setup_custom_swarm_and_client_proxy(
```
Comment (davidiw) [inline testsuite/tests/libratest/smoke_test.rs]: delete

File: testsuite/tests/libratest/smoke_test.rs [hunk 7]
```diff
@@ -201,7 +213,7 @@ fn test_smoke_script(mut client_proxy: ClientProxy) {
 
 #[test]
 fn test_execute_custom_module_and_script() {
-    let (_swarm, mut client_proxy) = setup_swarm_and_client_proxy(1, 0);
+    let (_swarm, mut client_proxy) = setup_swarm_and_client_proxy(1, 0, None);
```
Comment (davidiw) [inline testsuite/tests/libratest/smoke_test.rs]: this is the test that should have the open vm code

File: testsuite/tests/libratest/smoke_test.rs [hunk 8]
```diff
@@ -429,13 +441,14 @@ fn test_startup_sync_state() {
 
 #[test]
 fn test_basic_state_synchronization() {
-    //
```
Comment (davidiw) [inline testsuite/tests/libratest/smoke_test.rs]: nice

File: testsuite/tests/libratest/smoke_test.rs (hunk 1 above)
Comment (phoenix-antigravity) [reply testsuite/tests/libratest/smoke_test.rs]: hm, not sure why that specific default is weird though?
it's just ability to execute custom script. Doesn't have any implications on other tests and perf in general
would love to hear about more cases coming. Cause so far it looks like we could automatically cover multi-chunks in all related tests(e.g. restartability, full node sync, etc.) and all other call sites that keep passing None look weird

in short: a) I'd at least make this setting as default for all integration tests 
and b) would postpone adding complex signatures until there's real case. But don't have strong opinions on b), specially if you have more use cases in mind

File: libra-swarm/src/swarm.rs (hunk 4 above)
Comment (phoenix-antigravity) [reply libra-swarm/src/swarm.rs]: related to previous discussion, but I'd at least move it out of swarm.rs. Maybe to configuring defaults in smoke test itself

File: testsuite/tests/libratest/smoke_test.rs (hunk 1 above)
Comment (sunmilee) [reply testsuite/tests/libratest/smoke_test.rs]: sort of unrelated but out of curiosity, do we want to only test multi-chunk sync in all cases? assuming we'd want to test single chunk as well, although that would probably happen less in prod

Comment (phoenix-antigravity) [reply testsuite/tests/libratest/smoke_test.rs]: I'd say always multi-chunk. It should cover richer set of cases

Comment (davidiw) [reply testsuite/tests/libratest/smoke_test.rs]: weird was a poor choice of words, I should just say that they are configuring non-standard values. We ought to have smoke tests executing the default config unless there are reasons not to. You're right that this is isolated at this point in time, but I've seen the template all over the place and have not really seen any use cases except for this one. In light of the fact that we'd be moving to a default chunk of 2, I'm fine with making that the default within smoke test. So we can still benefit from @sunmilee's work at the interface between smoketests and libraswarm.

Comment (sunmilee) [reply testsuite/tests/libratest/smoke_test.rs]: Actually VMPublishingOption needs to be `VMPublishingOption::Open` just for `test_execute_custom_module_and_script` and doesn't need to be set as this for others. Should we make that as default too then?

Comment (phoenix-antigravity) [reply testsuite/tests/libratest/smoke_test.rs]: @davidiw: sorry for confusion. I have nothing against interface between smoketest and libraswarm. I just think signature of setup_swarm_and_client_proxy doesn't have to include template parameter yet. Cause IMO chunk size should be set globally and in all other places we just pass None for now. I'd add it once there's real need for it

Comment (sunmilee) [reply testsuite/tests/libratest/smoke_test.rs]: I'll keep the nodeconfig vs filepath refactor and make smoke test use the default chunk size instead of passing optional param.
